src/troubleshooting/content.md

```diff
@@ -0,0 +1,641 @@
+
+![Flow chart to debug deployments in Kubernetes](diagram.png)
+
+When you wish to deploy an application in Kubernetes, you usually define three components:
+
+- a Deployment — which is a recipe for creating copies of your application called Pods
+- a Service — an internal load balancer that routes the traffic to Pods
+- an Ingress — a description of how the traffic should flow from outside the cluster to your Service.
+
+Here's a quick visual recap.
+
+```slideshow
+{
+  "description": "Kubernetes core concepts: Ingress, Service, Pod and Deployment.",
+  "slides": [
+    {
+      "image": "core-concepts-1.svg",
+      "description": "In Kubernetes your applications are exposed through two layers of load balancers: internal and external."
+    },
+    {
+      "image": "core-concepts-2.svg",
+      "description": "The internal load balancer is called Service, whereas the external one is called Ingress."
+    },
+    {
+      "image": "core-concepts-3.svg",
+      "description": "Pods are not deployed directly. Instead, the Deployment creates the Pods and whatches over them."
+    }
+  ]
+}
+```
+
+Assuming you wish to deploy a simple _Hello World_ application, the YAML for such application should look similar to this:
+
+```yaml|highlight=2,23,34|title=hello-world.yaml
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: app
+  labels:
+    track: canary
+spec:
+  selector:
+    matchLabels:
+      name: app
+  template:
+    metadata:
+      labels:
+        name: app
+    spec:
+      containers:
+      - name: app
+        image: learnk8s/app:1.0.0
+        ports:
+        - containerPort: 8080
+---
+apiVersion: v1
+kind: Service
+metadata:
+  name: app
+spec:
+  ports:
+  - port: 80
+    targetPort: 8080
+  selector:
+    name: app
+---
+apiVersion: networking.k8s.io/v1beta1
+kind: Ingress
+metadata:
+  name: app
+spec:
+  rules:
+  - http:
+    paths:
+    - backend:
+        serviceName: app
+        servicePort: 80
+      path: /
+```
+
+The definition is quite long, and it's easy to overlook how the components relate to each other.
+
+For example:
+
+- When should you use port 80 and when port 8080?
+- Should you create a new port for every Service so that they don't clash?
+- Do label names matter? Should it be the same everywhere?
+
+Before focusing on the debugging, let's recap how the three components link to each other.
+
+Let's start with Deployment and Service.
+
+## Connecting Deployment and Service
+
+The surprising news is that Service and Deployment aren't connected at all.
+
+Instead, the Service points to the Pods directly and skips the Deployment altogether.
+
+So what you should pay attention to is how the Pods and the Service are related to each other.
+
+You should remember three things:
+
+1. The Service selector should match the label of the Pod
+1. The Service `targetPort` should match the `containerPort` of the container inside the Pod
```

Comment:
and the app have to issue a syscall to listen on that port, containerPort does nothing.

Reply:
Ok, but this guide is for someone who doesn't know syscall and has a broken deployment.
I could have written that `containerPort` is useless, but explaining it is a blog on its own?